### PR TITLE
Multi-Processing Functionality for create-workload

### LINKED ISSUE #375: Create Workload Improvement: Multi-Process Index Extraction
**Is your feature request related to a problem? Please describe.**

When using OpenSearch Benchmark's `create-workload` feature to extract data from clusters, users often extract more than one index. For example, a user might have a cluster comprised of 5 indices and OSB will extract each corpora in order. This is a time consuming process and can be more efficient. For larger sets of data, this can also lead to read timeouts like below:
```
[ERROR] Cannot create-workload. ConnectionTimeout caused by - ReadTimeoutError(HTTPSConnectionPool(host='<example host endpoint>', port=443): Read timed out. (read timeout=10)).
```

**Describe the solution you'd like**

To expedite the process and make it more efficient, OSB can use N number of processes to extract N number of indices. 

**Describe alternatives you've considered**

**Additional context**

### PR DESCRIPTION
**Pull request by AkshathRaghav**
### Description
Adding multi-process functionality that works concurrently. 

### Issues Resolved
#375 

### Testing
- [x] Tested if the create_workloads completes, and checks if it runs properly.

**REQUEST FOR HELP**: 
Below, is the sample output for the create_workloads with multi-processing. The render_progress is not working as expected, I have added a lock() to not overload and print for every thread thats running, but it doesn't seem to be working. Please help if you can! 

```
╭─aksha@Akshaths-PC ~/Workbench/opensearch-benchmark ‹AkshathRaghav/multi_process_index_extraction●› 
╰─$ opensearch-benchmark create-workload \
--workload=flights \
--target-host="https://127.0.0.1:9200" \
--client-options="basic_auth_user:'admin',basic_auth_password:'admin'" \
--output-path=~/Workbench/Workloads \
--indices=opensearch_dashboards_sample_data_flights \
--number-of-docs opensearch_dashboards_sample_data_flights:2500 \
--client-options="timeout:300,use_ssl:true,verify_certs:false,basic_auth_user:'admin',basic_auth_password:'admin'" \


   ____                  _____                      __       ____                  __                         __
  / __ \____  ___  ____ / ___/___  ____ ___________/ /_     / __ )___  ____  _____/ /_  ____ ___  ____ ______/ /__
 / / / / __ \/ _ \/ __ \\__ \/ _ \/ __ `/ ___/ ___/ __ \   / __  / _ \/ __ \/ ___/ __ \/ __ `__ \/ __ `/ ___/ //_/
/ /_/ / /_/ /  __/ / / /__/ /  __/ /_/ / /  / /__/ / / /  / /_/ /  __/ / / / /__/ / / / / / / / / /_/ / /  / ,<
\____/ .___/\___/_/ /_/____/\___/\__,_/_/   \___/_/ /_/  /_____/\___/_/ /_/\___/_/ /_/_/ /_/ /_/\__,_/_/  /_/|_|
    /_/

[INFO] Connected to OpenSearch cluster [opensearch-node1] version [2.11.0].

Extracting documents for index [opensearch_dashboards_sampl...     2/1000 docs [0.2% done]                       Extracting documents for index [opensearch_dashboards_sampl...     3/1000 docs [0.3% done]                       Extracting documents for index 
[opensearch_dashboards_sa...     251/1000 docs [25.1% done]                       Extracting documents for index [opensearch_dashboards_sa...     252/1000 docs [25.2% done]                       Extracting documents for index [opensearch_dashboards_sa...     627/1000 docs [62.7% done]                       Extracting documents for index [opensearch_dashboards_sa...     253/1000 docs [25.3% done]                       Extracting documents for index [opensearch_dashboards_sa...     501/1000 docs [50.1% done]                       Extracting documents for index [opensearch_dashboards_sa...     875/1000 docs [87.5% done]                       Extracting documents for index [opensearch_dashboards_sampl...     2/2500 docs [0.1% done]                       Extracting documents for index [opensearch_dashboards_sa...     626/2500 docs [25.0% done]                       Extracting documents for index [opensearch_dashboards_sampl...     4/2500 docs [0.2% done]                       Extracting documents for index [opensearch_dashboards_s...     1562/2500 docs [62.5% done]                       Extracting documents for index [opensearch_dashboards_s...     1874/2500 docs [75.0% done]                       Extracting documents for index [opensearch_dashboards_sa...     316/2500 docs [12.6% done]                       Extracting documents for index [opensearch_dashboards_s...     1250/2500 docs [50.0% done]                       Extracting documents for index [opensearch_dashboards_sa...     936/2500 docs [37.4% done]                       Extracting documents for index [opensearch_dashboards_s...     2496/2500 docs [99.8% done]
[INFO] Workload flights has been created. Run it with: opensearch-benchmark --workload-path=/home/aksha/Workbench/Workloads/flights

-------------------------------
[INFO] SUCCESS (took 5 seconds)
-------------------------------
```

---
By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: osbenchmark/workload_generator/corpus.py [hunk 1]
```diff
@@ -64,49 +69,162 @@ def extract(client, output_path, index, number_of_docs_requested=None):
...
 
 
-def dump_documents(client, index, out_path, number_of_docs, progress_message_suffix=""):
-    # pylint: disable=import-outside-toplevel
-    from opensearchpy import helpers
+def dump_documents_range(
+    client, index, out_path, start_doc, end_doc, total_docs, progress_message_suffix=""
+):
+    """
+    Extract documents in the range of start_doc and end_doc and write to induvidual files
```
Comment (IanHoang) [inline osbenchmark/workload_generator/corpus.py:107]: Nit: `induvidual files` --> `individual files`

File: osbenchmark/workload_generator/corpus.py [hunk 2]
```diff
@@ -64,49 +69,162 @@ def extract(client, output_path, index, number_of_docs_requested=None):
...
+
+    :param client: OpenSearch client used to extract data
+    :param index: Name of index to dump
```
Comment (IanHoang) [inline osbenchmark/workload_generator/corpus.py:110]: To help newcomers understand better, we can redefine this as:
`name of OpenSearch index to extract documents from`

**Comment by IanHoang** [comment]
Thanks for tackling this @AkshathRaghav! 

We should still offer users the option to go with the original route and not perform concurrent dumping. Adding a parameter like `--concurrent` could easily allow users to switch between the two. Also, does this feature work when users want to extract 2+ indices? If so, could you test that out and provide it in the description? It'd be nice to include the following comparisons in the description to highlight the performance benefits:
- single threaded extraction vs concurrent extraction of a single index
- single threaded extraction vs concurrent extraction of 2+ indices

Some additional notes:
1. For best practices, please provide a brief summary of your changes you're proposing within the PR description
2. Please revise the following lint errors detected by CI Unittests:
```
osbenchmark/workload_generator/corpus.py:128:12: W1203: Use lazy % formatting in logging functions (logging-fstring-interpolation)
osbenchmark/workload_generator/corpus.py:33:0: C0411: standard import "from threading import Lock" should be placed before "from osbenchmark.utils import console" (wrong-import-order)
```

File: osbenchmark/workload_generator/corpus.py [hunk 3]
```diff
@@ -64,49 +67,187 @@ def extract(client, output_path, index, number_of_docs_requested=None):
...
+    First, they split the documents into chunks to be dumped. Then, they are dumped as "{index}-documents{suffix}_{start}_{end}.json(.bz2)"
+    Finally, they are all collated into their file "{out_path}-documents{suffix}.json(.bz2)" format.
+
+    :param client: OpenSearch client used to extract data
+    :param index: Name of OpenSearch index to extract documents from
+    :param out_path: Destination directory for corpus dump
+    :param number_of_docs: Total number of documents
+    """
+    if concurrent:
+        num_threads = 8
```
Comment (gkamat) [inline osbenchmark/workload_generator/corpus.py:202]: The number of threads should be made configurable.

**Review by gkamat — CHANGES_REQUESTED** [review]

File: osbenchmark/workload_generator/corpus.py [hunk 4]
```diff
@@ -64,49 +67,187 @@ def extract(client, output_path, index, number_of_docs_requested=None):
...
+                end_doc,
+            )
+            query = {
+                "query": {"match_all": {}},
+                "from": start_doc,
+                "size": end_doc - start_doc,
+            }
+
+            batch_size = (end_doc - start_doc) // 5
+            search_after = None
```
Comment (IanHoang) [inline osbenchmark/workload_generator/corpus.py:146]: When will `search_after` not be None?

**Comment by gkamat** [comment]
@AkshathRaghav please also elaborate the title so it clear which component of OSB this is addressing.  Thanks

File: osbenchmark/workload_generator/corpus.py (hunk 4 above)
Comment (AkshathRaghav) [reply osbenchmark/workload_generator/corpus.py:146]: Like everytime the document has been retrieved, search_after gets set to that. This keeps going on until one batch is over. So in the next call for a batch, search_after becomes the (last document retreieved + 1)

**Comment by AkshathRaghav** [comment]
Some benchmarks: 


- Single threaded extraction vs concurrent extraction of a single index
   - Concurrent
```
╰─$ opensearch-benchmark create-workload \                    
--workload=flights \
--target-host="https://127.0.0.1:9200" \
--client-options="basic_auth_user:'admin',basic_auth_password:'admin'" \
--output-path=~/Workbench/Workloads \
--indices=opensearch_dashboards_sample_data_flights \
--number-of-docs=opensearch_dashboards_sample_data_flights:2500 \
--client-options="timeout:300,use_ssl:true,verify_certs:false,basic_auth_user:'admin',basic_auth_password:'admin'" \
> --concurrent

   ____                  _____                      __       ____                  __                         __
  / __ \____  ___  ____ / ___/___  ____ ___________/ /_     / __ )___  ____  _____/ /_  ____ ___  ____ ______/ /__
 / / / / __ \/ _ \/ __ \\__ \/ _ \/ __ `/ ___/ ___/ __ \   / __  / _ \/ __ \/ ___/ __ \/ __ `__ \/ __ `/ ___/ //_/
/ /_/ / /_/ /  __/ / / /__/ /  __/ /_/ / /  / /__/ / / /  / /_/ /  __/ / / / /__/ / / / / / / / / /_/ / /  / ,<
\____/ .___/\___/_/ /_/____/\___/\__,_/_/   \___/_/ /_/  /_____/\___/_/ /_/\___/_/ /_/_/ /_/ /_/\__,_/_/  /_/|_|
    /_/

[INFO] Connected to OpenSearch cluster [opensearch-node1] version [2.11.0].

Extracting documents from opensearch_dashboards_sample_data_flights [for test mode]: 100%|███████████████████████████████████████████████████████████████████████| 1000/1000 [00:00<00:00, 4939.46doc/s]
Extracting documents from opensearch_dashboards_sample_data_flights: 100%|███████████████████████████████████████████████████████████████████████████████████████| 2500/2500 [00:00<00:00, 6910.65doc/s]


[INFO] Workload flights has been created. Run it with: opensearch-benchmark --workload-path=/home/aksha/Workbench/Workloads/flights

-------------------------------
[INFO] SUCCESS (took 1 seconds)
-------------------------------
```
   - Single Threaded
```
╰─$ opensearch-benchmark create-workload \
--workload=flights \
--target-host="https://127.0.0.1:9200" \
--client-options="basic_auth_user:'admin',basic_auth_password:'admin'" \
--output-path=~/Workbench/Workloads \
--indices=opensearch_dashboards_sample_data_flights \                                            
--number-of-docs=opensearch_dashboards_sample_data_flights:2500 \                                                 
--client-options="timeout:300,use_ssl:true,verify_certs:false,basic_auth_user:'admin',basic_auth_password:'admin'"

   ____                  _____                      __       ____                  __                         __
  / __ \____  ___  ____ / ___/___  ____ ___________/ /_     / __ )___  ____  _____/ /_  ____ ___  ____ ______/ /__
 / / / / __ \/ _ \/ __ \\__ \/ _ \/ __ `/ ___/ ___/ __ \   / __  / _ \/ __ \/ ___/ __ \/ __ `__ \/ __ `/ ___/ //_/
/ /_/ / /_/ /  __/ / / /__/ /  __/ /_/ / /  / /__/ / / /  / /_/ /  __/ / / / /__/ / / / / / / / / /_/ / /  / ,<
\____/ .___/\___/_/ /_/____/\___/\__,_/_/   \___/_/ /_/  /_____/\___/_/ /_/\___/_/ /_/_/ /_/ /_/\__,_/_/  /_/|_|
    /_/

[INFO] Connected to OpenSearch cluster [opensearch-node1] version [2.11.0].

Extracting documents from opensearch_dashboards_sample_data_flights [for test mode]: 100%|███████████████████████████████████████████████████████████████████████| 1000/1000 [00:00<00:00, 4484.50doc/s]
Extracting documents from opensearch_dashboards_sample_data_flights: 100%|███████████████████████████████████████████████████████████████████████████████████████| 2500/2500 [00:00<00:00, 4878.57doc/s]


[INFO] Workload flights has been created. Run it with: opensearch-benchmark --workload-path=/home/aksha/Workbench/Workloads/flights

-------------------------------
[INFO] SUCCESS (took 1 seconds)
-------------------------------
```
- Single threaded extraction vs concurrent extraction of 2+ indices
  - Concurrent
```
╰─$ opensearch-benchmark create-workload \
--workload=flights \
--target-host="https://127.0.0.1:9200" \
--client-options="basic_auth_user:'admin',basic_auth_password:'admin'" \
--output-path=~/Workbench/Workloads \
--indices=opensearch_dashboards_sample_data_flights,opensearch_dashboards_sample_data_ecommerce \
--number-of-docs=opensearch_dashboards_sample_data_flights:2500,opensearch_dashboards_sample_data_ecommerce:1500 \
--client-options="timeout:300,use_ssl:true,verify_certs:false,basic_auth_user:'admin',basic_auth_password:'admin'" \
--concurrent

   ____                  _____                      __       ____                  __                         __
  / __ \____  ___  ____ / ___/___  ____ ___________/ /_     / __ )___  ____  _____/ /_  ____ ___  ____ ______/ /__
 / / / / __ \/ _ \/ __ \\__ \/ _ \/ __ `/ ___/ ___/ __ \   / __  / _ \/ __ \/ ___/ __ \/ __ `__ \/ __ `/ ___/ //_/
/ /_/ / /_/ /  __/ / / /__/ /  __/ /_/ / /  / /__/ / / /  / /_/ /  __/ / / / /__/ / / / / / / / / /_/ / /  / ,<
\____/ .___/\___/_/ /_/____/\___/\__,_/_/   \___/_/ /_/  /_____/\___/_/ /_/\___/_/ /_/_/ /_/ /_/\__,_/_/  /_/|_|
    /_/

[INFO] Connected to OpenSearch cluster [opensearch-node1] version [2.11.0].

Extracting documents from opensearch_dashboards_sample_data_flights [for test mode]: 100%|███████████████████████████████████████████████████████████████████████| 1000/1000 [00:00<00:00, 4673.39doc/s]
Extracting documents from opensearch_dashboards_sample_data_flights: 100%|███████████████████████████████████████████████████████████████████████████████████████| 2500/2500 [00:00<00:00, 6522.05doc/s]
Extracting documents from opensearch_dashboards_sample_data_ecommerce [for test mode]: 100%|█████████████████████████████████████████████████████████████████████| 1000/1000 [00:00<00:00, 3324.74doc/s]
Extracting documents from opensearch_dashboards_sample_data_ecommerce: 100%|█████████████████████████████████████████████████████████████████████████████████████| 1500/1500 [00:00<00:00, 3821.05doc/s]


[INFO] Workload flights has been created. Run it with: opensearch-benchmark --workload-path=/home/aksha/Workbench/Workloads/flights

-------------------------------
[INFO] SUCCESS (took 2 seconds)
-------------------------------
```
   - Single Threaded
```
╰─$ opensearch-benchmark create-workload \
--workload=flights \
--target-host="https://127.0.0.1:9200" \
--client-options="basic_auth_user:'admin',basic_auth_password:'admin'" \
--output-path=~/Workbench/Workloads \
--indices=opensearch_dashboards_sample_data_flights,opensearch_dashboards_sample_data_ecommerce \
--number-of-docs=opensearch_dashboards_sample_data_flights:2500,opensearch_dashboards_sample_data_ecommerce:1500 \
--client-options="timeout:300,use_ssl:true,verify_certs:false,basic_auth_user:'admin',basic_auth_password:'admin'"  

   ____                  _____                      __       ____                  __                         __
  / __ \____  ___  ____ / ___/___  ____ ___________/ /_     / __ )___  ____  _____/ /_  ____ ___  ____ ______/ /__
 / / / / __ \/ _ \/ __ \\__ \/ _ \/ __ `/ ___/ ___/ __ \   / __  / _ \/ __ \/ ___/ __ \/ __ `__ \/ __ `/ ___/ //_/
/ /_/ / /_/ /  __/ / / /__/ /  __/ /_/ / /  / /__/ / / /  / /_/ /  __/ / / / /__/ / / / / / / / / /_/ / /  / ,<
\____/ .___/\___/_/ /_/____/\___/\__,_/_/   \___/_/ /_/  /_____/\___/_/ /_/\___/_/ /_/_/ /_/ /_/\__,_/_/  /_/|_|
    /_/

[INFO] Connected to OpenSearch cluster [opensearch-node1] version [2.11.0].

Extracting documents from opensearch_dashboards_sample_data_flights [for test mode]: 100%|███████████████████████████████████████████████████████████████████████| 1000/1000 [00:00<00:00, 4572.21doc/s]
Extracting documents from opensearch_dashboards_sample_data_flights: 100%|███████████████████████████████████████████████████████████████████████████████████████| 2500/2500 [00:00<00:00, 4899.36doc/s]
Extracting documents from opensearch_dashboards_sample_data_ecommerce [for test mode]: 100%|█████████████████████████████████████████████████████████████████████| 1000/1000 [00:00<00:00, 2700.82doc/s]
Extracting documents from opensearch_dashboards_sample_data_ecommerce: 100%|█████████████████████████████████████████████████████████████████████████████████████| 1500/1500 [00:00<00:00, 2632.83doc/s]


[INFO] Workload flights has been created. Run it with: opensearch-benchmark --workload-path=/home/aksha/Workbench/Workloads/flights

-------------------------------
[INFO] SUCCESS (took 2 seconds)
-------------------------------
```


All these methods run in under a second, so instead I've been comparing them with respect to extracting+dumping speed which is overall faster in the concurrent runs. 

@gkamat I am working on making some parameters more configurable right now!

**Comment by AkshathRaghav** [comment]
@gkamat I've made the requested changes. Currently working on changing unit tests. 

```
╰─$ opensearch-benchmark create-workload --workload=flights --target-host="https://127.0.0.1:9200" \
--client-options="basic_auth_user:'admin',basic_auth_password:'admin'" \
--output-path=~/Workbench/Workloads \
--indices=opensearch_dashboards_sample_data_flights,opensearch_dashboards_sample_data_ecommerce \
--number-of-docs=opensearch_dashboards_sample_data_flights:2500,opensearch_dashboards_sample_data_ecommerce:1500 \
--client-options="timeout:300,use_ssl:true,verify_certs:false,basic_auth_user:'admin',basic_auth_password:'admin'" \
--concurrent --threads=8 --bsize=50 --custom_dump_query=../test/custom_query.json


   ____                  _____                      __       ____                  __                         __
  / __ \____  ___  ____ / ___/___  ____ ___________/ /_     / __ )___  ____  _____/ /_  ____ ___  ____ ______/ /__
 / / / / __ \/ _ \/ __ \\__ \/ _ \/ __ `/ ___/ ___/ __ \   / __  / _ \/ __ \/ ___/ __ \/ __ `__ \/ __ `/ ___/ //_/
/ /_/ / /_/ /  __/ / / /__/ /  __/ /_/ / /  / /__/ / / /  / /_/ /  __/ / / / /__/ / / / / / / / / /_/ / /  / ,<
\____/ .___/\___/_/ /_/____/\___/\__,_/_/   \___/_/ /_/  /_____/\___/_/ /_/\___/_/ /_/_/ /_/ /_/\__,_/_/  /_/|_|
    /_/

[INFO] Connected to OpenSearch cluster [opensearch-node1] version [2.11.0].

Extracting documents from opensearch_dashboards_sample_data_flights [for test mode]: 100%|███████████████████████████████████████████████████████████████████████████████████████| 1000/1000 [00:00<00:00, 6777.74doc/s]
Extracting documents from opensearch_dashboards_sample_data_flights: 100%|███████████████████████████████████████████████████████████████████████████████████████████████████████| 2500/2500 [00:00<00:00, 9417.50doc/s]
Extracting documents from opensearch_dashboards_sample_data_ecommerce [for test mode]: 100%|█████████████████████████████████████████████████████████████████████████████████████| 1000/1000 [00:00<00:00, 5065.93doc/s]
Extracting documents from opensearch_dashboards_sample_data_ecommerce: 100%|█████████████████████████████████████████████████████████████████████████████████████████████████████| 1500/1500 [00:00<00:00, 5928.33doc/s]


[INFO] Workload flights has been created. Run it with: opensearch-benchmark --workload-path=/home/aksha/Workbench/Workloads/flights

-------------------------------
[INFO] SUCCESS (took 1 seconds)
-------------------------------
```

**Comment by AkshathRaghav** [comment]
@gkamat @IanHoang Hi! I think I've done most of what is required for this PR. Please let me know if everything is ok. 
I've changed unit tests such that make it and make test work. I've checked the linting as well. 

I tried adding the commit from the conflicting file, but security check still shows it. Not sure what to do on that front. 
Another question I had was that the checks are not running for some reason... I'm not sure where I went wrong. Please let me know what I have to do to solve that!

**Comment by IanHoang** [comment]
> @gkamat @IanHoang Hi! I think I've done most of what is required for this PR. Please let me know if everything is ok. I've changed unit tests such that make it and make test work. I've checked the linting as well.
> 
> I tried adding the commit from the conflicting file, but security check still shows it. Not sure what to do on that front. Another question I had was that the checks are not running for some reason... I'm not sure where I went wrong. Please let me know what I have to do to solve that!

The checks aren't running since there are conflicts in benchmark.py. Perhaps try resolving conflicts from the latest changes on mainline with this branch?

File: osbenchmark/benchmark.py [hunk 5]
```diff
@@ -139,115 +181,161 @@ def add_workload_source(subparser):
...
+        help="Number of threads for dumping documents from indices (default: false)",
+    )
+    create_workload_parser.add_argument(
+        "--bsize",
+        type=positive_number,
+        default=0,  # 0 means dynamic batching within corpus.py `dump_documents_range()`
+        help="Batch size to use for dumping documents (default: false)",
+    )
+    create_workload_parser.add_argument(
+        "--custom_dump_query",
```
Comment (IanHoang) [inline osbenchmark/benchmark.py:259]: --custom_dump_query --> --custom-dump-query

File: osbenchmark/benchmark.py [hunk 6]
```diff
@@ -139,115 +181,161 @@ def add_workload_source(subparser):
...
+        help="Whether to generate workload concurrently (default: false)",
+    )
+    create_workload_parser.add_argument(
+        "--threads",
+        type=positive_number,
+        default=8,
+        help="Number of threads for dumping documents from indices (default: false)",
+    )
+    create_workload_parser.add_argument(
+        "--bsize",
```
Comment (IanHoang) [inline osbenchmark/benchmark.py:253]: Nit: I think it's worth changing this to `--batch-size` to make it easier to understand on first glance and to keep it consistent with every other parameter that spells out the option. One thing we can do in the future is add an abbreviated option for specific parameters. For example, users can either use `--batch-size` or `-b`.

File: osbenchmark/benchmark.py [hunk 7]
```diff
@@ -139,115 +181,161 @@ def add_workload_source(subparser):
...
+    create_workload_parser.add_argument(
+        "--concurrent",
+        action="store_true",
+        help="Whether to generate workload concurrently (default: false)",
+    )
+    create_workload_parser.add_argument(
+        "--threads",
+        type=positive_number,
+        default=8,
+        help="Number of threads for dumping documents from indices (default: false)",
```
Comment (IanHoang) [inline osbenchmark/benchmark.py:250]: To clarify, we should elaborate the help statement:
```
Number of threads for dumping documents from indices. To use this parameter, command must include --concurrent parameter. By default, this parameter is disabled. When this parameter is enabled, number of threads defaults to 8
```

**Pull request closed without merging**